numalogic/models/threshold/_mahalanobis.py

```diff
@@ -0,0 +1,168 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Final
+
+import numpy as np
+import numpy.typing as npt
+
+from numalogic.base import BaseThresholdModel
+from typing_extensions import Self
+
+from numalogic.tools.exceptions import ModelInitializationError
+
+_INLIER: Final[int] = 0
+_OUTLIER: Final[int] = 1
+
+
+class MahalanobisThreshold(BaseThresholdModel):
+    """
+    Multivariate threshold estimator using Mahalanobis distance.
+
+    The maximum outlier probability is used to calculate the k value
+    using Chebyshev's inequality. This basically means that the
+    probability of values lying outside the interval
+    (mean - k * std, mean + k * std) is not more than max_outlier_prob.
+    A value of 0.1 means that 90% of the values lie within the interval and
+    10% of the values lie outside the interval.
+
+    The threshold is calculated as the
+    mean of Mahalanobis distance plus k times the standard deviation
+    of Mahalanobis distance.
+
+    Args:
+    ----
+        max_outlier_prob: maximum outlier probability (default: 0.1)
+
+    Raises
+    ------
+        ValueError: if max_outlier_prob is not in range (0, 1)
+
+    >>> import numpy as np
+    >>> from numalogic.models.threshold import MahalanobisThreshold
+    >>> rng = np.random.default_rng(42)
+    ...
+    >>> x_train = rng.normal(size=(100, 15))
+    >>> x_test = rng.normal(size=(30, 15))
+    ...
+    >>> clf = MahalanobisThreshold()
+    >>> clf.fit(x_train)
+    ...
+    >>> y_test = clf.predict(x_test)
+    >>> test_scores = clf.score_samples(x_test)
+    """
+
+    def __init__(self, max_outlier_prob: float = 0.1):
+        if not 0.0 < max_outlier_prob < 1.0:
+            raise ValueError("max_outlier_prob should be in range (0, 1)")
+        self._k = self._chebyshev_k(max_outlier_prob)
+        self._distr_mean = None
+        self._cov_inv = None
+        self._md_thresh = None
+        self._is_fitted = False
+
+    @property
+    def threshold(self) -> float:
+        """Returns the threshold value."""
+        return self._md_thresh
+
+    @property
+    def std_factor(self) -> float:
+        """Returns the k value calculated using Chebyshev's inequality."""
+        return self._k
+
+    @staticmethod
+    def _chebyshev_k(p: float) -> float:
+        """Calculate the k value using Chebyshev's inequality."""
+        return np.reciprocal(np.sqrt(p))
+
+    def fit(self, x: npt.NDArray[float]) -> Self:
+        """
+        Fit the estimator on the training set.
+
+        Args:
+        ----
+            x: training data of shape (n_samples, n_features)
+
+        Returns
+        -------
+            self
+        """
+        self._distr_mean = np.mean(x, axis=0)
+        cov = np.cov(x, rowvar=False)
```

Comment:
_A major issue with the MD is that the inverse of the correlation matrix is needed for the calculations. This can’t be calculated if the variables are highly correlated (Varmuza & Filzmoser, 2016)._

Shall we put a check for this?

Reply:
If the variables are highly correlated, then basically it means that the matrix has a chance of being singular in nature. We use the Moore-Penrose Pseudo inverse of a matrix rather than an actual inverse, which mitigates this issue somewhat.

Reply:
Oh.....just saw your psuedo inverse change. Interesting thing!
Resolving this.